docs/writing-intel-modules.md

```diff
@@ -0,0 +1,248 @@
+
+This doc contains guidelines on creating a Cartography intel module.  If you want to add a new data type to Cartography,
+this is the guide for you!  It is fairly straightforward to copy the structure of an existing intel module and test it,
+but we'll share some best practices in this doc to save you some time.  We look forward to receiving your PR!
+
+## Before getting started...
+
+Read through and follow the setup steps in [the Cartography developer guide](developer-guide.md).  Learn the basics of
+running, testing, and linting your code there.
+
+
+## Sync = Get, Transform, Load, Cleanup
+
+A cartography intel module consists of one `sync` function.  `sync` should call `get`, then `load`, and finally `cleanup`.
+
+
+### Get
+
+The `get` function [retrieves necessary data](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/intel/gcp/compute.py#L98)
+from a resource provider API, which is GCP in this particular example.
+
+`get` should be "dumb" in the sense that it should not handle retry logic or data
+manipulation.
+
+### Transform
+
+The `transform` function [manipulates data](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/intel/gcp/compute.py#L193)
+to make it easier to ingest to the graph.  We have some best practices on handling transforms:
+
+#### Handling required versus optional fields
+
+TL;DR: Use `my_dict['someField']` when accessing important fields, and use `my_dict.get('someField')` for less important ones.
+
+We should directly access dicts in cases where not having the data should cause a sync to fail.
+For example, if we are transforming AWS data, we definitely need an AWS object's ARN field because it uniquely
+identifies the object.  Therefore, we should access an object's ARN using `data['arn']` as opposed to
+using `data.get('arn')` (the former will raise a `KeyError` if `arn` does not exist and the latter will just return
+`None` without an exception).
+
+We _want_ the sync to fail if an important field is not present in our data.  The idea here is that
+it is better to fail a sync than to add malformed data.
+
+On the other hand, we should use `data.get('SomeField')` if `SomeField` is something optional that can afford to be
+`None`.
+
+
+### Load
+
+The `load` function ingests the processed data to Neo4j, [as seen in this GCP VPC example](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/intel/gcp/compute.py#L442).
+There are many best practices to consider here.
+
+#### Handling cartography's `update_tag`:
+
+`cartography`'s global [config object carries around an `update_tag` property](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/cli.py#L91-L98) which is
+[set to the time that the CLI is run](https://github.com/lyft/cartography/blob/8d60311a10156cd8aa16de7e1fe3e109cc3eca0f/cartography/sync.py#L131-L134).
```

Comment:
More properly, it's a tag/label associated with the current sync. The CLI sets this to the UNIX timestamp representing the time of invocation, but in theory it could be anything -- provided it is unique per sync. Don't think we need to go into a huge amount of detail here, but do think it's important to be a bit more abstract.